ui/src/testing/stubs/workspaces-api-stub.ts

```diff
@@ -1,13 +1,59 @@
 import {
+  FileDetail,
   WorkspacesApi
 } from 'generated/fetch';
 
 export class WorkspacesApiStub extends WorkspacesApi {
+  notebookList: FileDetail[];
+
   constructor() {
     super(undefined, undefined, (..._: any[]) => { throw Error('cannot fetch in tests'); });
+
+    this.notebookList = WorkspacesApiStub.stubNotebookList();
   }
 
-  getNoteBookList() {
-    return Promise.resolve([]);
+  static stubNotebookList(): FileDetail[] {
+    return [
+      {
+        'name': 'mockFile.ipynb',
+        'path': 'gs://bucket/notebooks/mockFile.ipynb',
+        'lastModifiedTime': 100
+      }
+    ];
   }
+
+  getNoteBookList(workspaceNamespace: string,
+    workspaceId: string, extraHttpRequestParams?: any): Promise<Array<FileDetail>> {
+    return new Promise<Array<FileDetail>>(resolve => {
+      setTimeout(() => {
+        resolve(this.notebookList);
+      }, 0);
+    });
+  }
+
+  cloneNotebook(workspaceNamespace: string, workspaceId: string,
+    notebookName: String): Promise<any> {
+    return new Promise<any>(resolve => {
+      setTimeout(() => {
+        const cloneName = notebookName.replace('.ipynb', '') + ' Clone.ipynb';
+        this.notebookList.push({
+          'name': cloneName,
+          'path': 'gs://bucket/notebooks/' + cloneName,
+          'lastModifiedTime': 100
+        });
+        resolve({});
```

Comment:
Do we really intend to implement a whole fake backend for the sake of tests? What's actually being verified here?

Reply:
For behavioral tests (which are generally what I would favor in terms of value), having a single fake service implementation tends to be more maintainable and cleaner than setting up complicated multistage and implementation-dependent mock expectations.

e.g. test which does a notebook clone on the list page:

```
// mocks setup

mock(clone(...));
mock(listNotebooks(...)).return(fakePostCloneNotebooks)
// click clone
// verify UI shows cloned card
```

vs

```
// stub setup

// click clone
// check stub service or UI for cloned notebook
```

Note in the first case, that the actual details of what clone "does" have been simulated by the mocks (in each test case that needs it) - vs just implementing the logic in a fake. Using fakes also tends to scale a bit better in terms of readability in my experience especially when there are multiple services involved.

Anyways - this is my experience, but the value we see here will entirely depend on the types of tests we are writing.